### PR TITLE
feat(api): make modules aware of available fw updates in file system

### PR DESCRIPTION
**Pull request by b-cooper**
## overview

Creates a generalized approach for determining whether an attached module is capable of being
updated, based on the module firmware artifacts bundled in the filesystem. In addition to adding
this functionality, it includes a couple of refactors to dedupe common module subclass
responsibilities. In debugging this branch, I also included some loop level error handling so
exceptions in asyncio tasks (e.g. module instance creation) don't get swallowed.

re #4575

### CODE REVIEW

File: api/src/opentrons/hardware_control/__init__.py [hunk 1]
```diff
@@ -44,6 +44,19 @@ def _log_call_inner(*args, **kwargs):
...
 
+def handle_loop_exception(loop: asyncio.AbstractEventLoop, context):
+    msg = context.get("exception", context["message"])
+    mod_log.error(f"Caught exception: {msg}")
+
+
+def use_or_initialize_loop(loop: Optional[asyncio.AbstractEventLoop]):
+    checked_loop = loop or asyncio.get_event_loop()
+
+    # TODO: BC 2020-01-24 use loop.add_signal_handler for proper cleanup
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/__init__.py]: i give this a big maybe - signal handlers have to be added from the main thread every time, and some signals can't be handled anyway.

File: api/src/opentrons/hardware_control/__init__.py [hunk 2]
```diff
@@ -44,6 +44,19 @@ def _log_call_inner(*args, **kwargs):
     return _log_call_inner
 
 
+def handle_loop_exception(loop: asyncio.AbstractEventLoop, context):
+    msg = context.get("exception", context["message"])
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/__init__.py]: Bit of a nitpick but we should really log both the exception and the message. Per the docs, neither is optional so we shouldn't have to use get here.

File: api/src/opentrons/hardware_control/__init__.py [hunk 3]
```diff
@@ -44,6 +44,19 @@ def _log_call_inner(*args, **kwargs):
     return _log_call_inner
 
 
+def handle_loop_exception(loop: asyncio.AbstractEventLoop, context):
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/__init__.py]: `context` could probably be typed as `Dict[str, Any]`

File: api/src/opentrons/hardware_control/__init__.py [hunk 4]
```diff
@@ -44,6 +44,19 @@ def _log_call_inner(*args, **kwargs):
...
+    msg = context.get("exception", context["message"])
+    mod_log.error(f"Caught exception: {msg}")
+
+
+def use_or_initialize_loop(loop: Optional[asyncio.AbstractEventLoop]):
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/__init__.py]: Annotate return type as `asyncio.AbstractEventLoop`

File: api/src/opentrons/hardware_control/modules/mod_abc.py [hunk 5]
```diff
@@ -13,15 +22,61 @@ class AbstractModule(abc.ABC):
...
         """
         pass
 
+    @abc.abstractmethod
+    def __init__(self,
+                 port: str,
+                 simulating: bool = False,
+                 loop: asyncio.AbstractEventLoop = None) -> None:
+        self._port = port
+        if None is loop:
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/mod_abc.py]: can use that helper function here

File: api/src/opentrons/hardware_control/modules/mod_abc.py [hunk 6]
```diff
@@ -13,15 +22,61 @@ class AbstractModule(abc.ABC):
...
+            self._loop = asyncio.get_event_loop()
+        else:
+            self._loop = loop
+        self._device_info = None
+        self._available_update_version: Optional[str] = None
+        self._available_update_path = self.get_bundled_fw()
+        return None
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/mod_abc.py]: kinda bad form, can omit

File: api/src/opentrons/hardware_control/modules/mod_abc.py [hunk 7]
```diff
@@ -13,15 +22,61 @@ class AbstractModule(abc.ABC):
...
+
+    def get_bundled_fw(self) -> Optional[Path]:
+        """ Get absolute path to bundled version of module fw if available. """
+        name_to_fw_file_prefix = {
+            "tempdeck": "temperature-module", "magdeck": "magnetic-module"}
+        name = self.name()
+        file_prefix = name_to_fw_file_prefix.get(name, name)
+        MODULE_FW_RE = re.compile(f'{file_prefix}@v(.*).(hex|bin)')
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/mod_abc.py]:
```suggestion
        MODULE_FW_RE = re.compile(f'{file_prefix}@v(.*)\.(hex|bin)')
```

Also do we really handle either hex or bin? IMO we should pick one

File: api/src/opentrons/hardware_control/modules/mod_abc.py [hunk 8]
```diff
@@ -13,15 +22,61 @@ class AbstractModule(abc.ABC):
...
+        fw_dir = CONFIG['robot_firmware_dir']
+        fw_resources = [fw_dir / item for item in os.listdir(fw_dir)]
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/mod_abc.py]:
```suggestion
        fw_resources = fw_dir.iterdir()
```

Path objects have this handy operation that returns an iterator, which we then iterate through in line 60 (could even combine the two now)

File: api/src/opentrons/hardware_control/modules/mod_abc.py [hunk 9]
```diff
@@ -13,15 +22,61 @@ class AbstractModule(abc.ABC):
...
+        for fw_resource in fw_resources:
+            matches = MODULE_FW_RE.search(str(fw_resource))
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/mod_abc.py]: We could and probably should do a `match` on  `fw_resource.name`

File: api/src/opentrons/hardware_control/modules/mod_abc.py [hunk 10]
```diff
@@ -13,15 +22,61 @@ class AbstractModule(abc.ABC):
...
+                self._available_update_version = matches.group(1)
+                return fw_resource
+
+        mod_log.info(f"no available fw file found for: {file_prefix}")
+        return None
+
+    def has_available_update(self) -> bool:
+        """ Return whether a newer firmware file is available """
+        if self._device_info is not None:
+            raw_device_version = self._device_info.get('version', None)
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/mod_abc.py]: In what conditions do we expect device_info to be valid but not have a version? Also we can just do `.get('version')`, returning None on key miss is implicit

File: api/src/opentrons/hardware_control/modules/mod_abc.py [hunk 11]
```diff
@@ -13,15 +22,61 @@ class AbstractModule(abc.ABC):
...
+                 port: str,
+                 simulating: bool = False,
+                 loop: asyncio.AbstractEventLoop = None) -> None:
+        self._port = port
+        if None is loop:
+            self._loop = asyncio.get_event_loop()
+        else:
+            self._loop = loop
+        self._device_info = None
+        self._available_update_version: Optional[str] = None
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/mod_abc.py]: IMO we should bundle these two into a named tuple or something and have a function that returns `Optional[ModuleFirmware]` so we can't possibly be in a one-but-not-the-other situation

File: api/src/opentrons/main.py [hunk 12]
```diff
@@ -36,8 +36,8 @@ def _find_smoothie_file():
     # Search for smoothie files in /usr/lib/firmware first then fall back to
     # value packed in wheel
     if IS_ROBOT:
-        resources.extend([ROBOT_FIRMWARE_DIR / item
-                          for item in os.listdir(ROBOT_FIRMWARE_DIR)])
+        fw_dir = CONFIG['robot_firmware_dir']
+        resources.extend([fw_dir / item for item in os.listdir(fw_dir)])
```
Comment (sfoster1) [inline api/src/opentrons/main.py]: can be `resources.extend(fw_dir.iterdir())` (or `resources.extend(list(fw_dir.iterdir()))` if extend doesn't iterate)

File: api/src/opentrons/hardware_control/modules/thermocycler.py [hunk 13]
```diff
@@ -402,6 +395,10 @@ def set_loop(self, newLoop):
     def port(self):
         return self._port
 
+    @property
+    def available_update_path(self):
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/modules/thermocycler.py]: needs a return value annotation

File: api/src/opentrons/hardware_control/modules/mod_abc.py (hunk 7 above)
Comment (sfoster1) [reply api/src/opentrons/hardware_control/modules/mod_abc.py]: And as mentioned below we should make this a full string match and search the filename

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Early thoughts on the draft. Structurally it looks ok but there's some inline comments

File: api/src/opentrons/hardware_control/modules/mod_abc.py (hunk 8 above)
Comment (b-cooper) [reply api/src/opentrons/hardware_control/modules/mod_abc.py]: so handy ✋

**Review by sfoster1 — APPROVED** [review]
Looking really good!

**Pull request merged**